File: lib/c_mini/src/litmus_header.ml
(* The Automagic Compiler Tormentor

   Copyright (c) 2018--2019 Matt Windsor and contributors

   ACT itself is licensed under the MIT License. See the LICENSE file in the
   project root for more information.

   ACT is based in part on code from the Herdtools7 project
   (https://github.com/herd/herdtools7) : see the LICENSE.herd file in the
   project root for more information. *)

open Base

module J :
  Plumbing.Jsonable_types.S with type t = Constant.t Act_litmus.Header.t =
Act_litmus.Header.Json (struct
  include Constant

  let parse_post_string (s : string) :
      Constant.t Act_litmus.Postcondition.t Or_error.t =
    Or_error.(
      s |> Act_c_lang.Frontend.Litmus_post.load_from_string
      >>= Act_litmus.Postcondition.With_errors.map_right_m
            ~f:Convert.constant)
end)

include J
include Plumbing.Loadable.Of_jsonable (J)
include Plumbing.Storable.Of_jsonable (J)

let equal = Act_litmus.Header.equal Constant.equal

module Filters = struct
  module Dump = Plumbing.Filter.Make (struct
    let name = "dump-header"

    type aux_i = unit

    type aux_o = unit

    let run (ctx : aux_i Plumbing.Filter_context.t)
        (ic : Stdio.In_channel.t) (oc : Stdio.Out_channel.t) :
        aux_o Or_error.t =
      Or_error.Let_syntax.(
        let%bind test =
          Frontend.load_from_ic ic
            ~path:(Plumbing.Filter_context.input_path_string ctx)
        in
        store_to_oc ~dest:oc (Litmus.Test.header test))
  end)

  module Replace = Plumbing.Filter.Make (struct
    let name = "replace-header"

    type aux_i = t

    type aux_o = unit

    let run (ctx : aux_i Plumbing.Filter_context.t)
        (ic : Stdio.In_channel.t) (oc : Stdio.Out_channel.t) :
        aux_o Or_error.t =
      let header = Plumbing.Filter_context.aux ctx in
      Or_error.Let_syntax.(
        let%bind test =
          Frontend.load_from_ic ic
            ~path:(Plumbing.Filter_context.input_path_string ctx)
        in
        let%map test' =
          Litmus.Test.try_map_header test ~f:(fun _ ->
              Or_error.return header)
        in
        Litmus.Pp.print oc test')
  end)
end

File: scripts/act_py/litmus_header.py
# The Automagic Compiler Tormentor
#
# Copyright (c) 2018--2019 Matt Windsor and contributors
#
# ACT itself is licensed under the MIT License. See the LICENSE file in the
# project root for more information.
#
# ACT is based in part on code from the Herdtools7 project
# (https://github.com/herd/herdtools7) : see the LICENSE.herd file in the
# project root for more information.

import json
import typing
from dataclasses import dataclass

from act_py import json_utils, litmus_id


@dataclass
class LitmusHeader:
    """The bit of a Litmus test that isn't code."""

    name: str
    locations: typing.Optional[typing.List[str]]
    init: typing.Optional[typing.Dict[str, int]]
    postcondition: typing.Optional[str]

    def rewrite_locals(self, rewriter: typing.Callable[[litmus_id.Lid], str]):
        if self.postcondition is not None:
            self.postcondition = litmus_id.rewrite_post_locals(
                self.postcondition, rewriter
            )

    def dump(self, fp: typing.TextIO) -> None:
        """Dumps this aux record to a file.

        :param fp: A file-like object to use as the target of the write.
        :return: Nothing.
        """
        json_utils.dump_dataclass(self, fp)


def of_dict(aux_dict: typing.Dict[str, typing.Any]) -> LitmusHeader:
    name = json_utils.str_field(aux_dict, "name")
    locations = aux_dict["locations"]
    init = aux_dict["init"]
    postcondition = aux_dict["postcondition"]
    return LitmusHeader(name=name, locations=locations, init=init, postcondition=postcondition)


def load(fp: typing.TextIO) -> LitmusHeader:
    """Loads an aux file from a file pointer.

    :param fp: The (text) file pointer from which we will load an aux file.
    :return: The resulting `Aux` object.
    """
    aux_dict: typing.Dict[str, typing.Any] = json.load(fp)
    return of_dict(aux_dict)
